### PR TITLE
types/result, util/lineiter: add package for a result type, use it

### PR DESCRIPTION
**Pull request by bradfitz**
This adds a new generic result type (motivated by golang/go#70084) to
try it out, and uses it in the new lineutil package (replacing the old
lineread package), changing that package to return iterators:
sometimes over []byte (when the input is all in memory), but sometimes
iterators over results of []byte, if errors might happen at runtime.

Updates #12912
Updates golang/go#70084

### CODE REVIEW

**Comment by bradfitz** [comment]
Thoughts? I wrote it to see how I feel about it, but I don't totally hate it at least.

File: types/result/result.go [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+package result
+
+// Of is either a T value or an error.
+//
+// Think of it like Rust or Swift's result types.
+// It's named "Of" because the fully qualified name
+// for callers reads result.Of[T].
+type Of[T any] struct {
+	v   T // valid if Err is nil; invalid if Err is non-nil
+	Err error
```
Comment (maisem) [inline types/result/result.go:15]: Why is this exported?

**Review by maisem — APPROVED** [review]

File: util/lineread/lineread.go [hunk 2]
```diff
@@ -6,19 +6,52 @@ package lineread
...
+		}
+		defer f.Close()
+		bs := bufio.NewScanner(f)
+		for bs.Scan() {
+			if !yield(result.Value(bs.Bytes())) {
+				return
+			}
+		}
+		if err := bs.Err(); err != nil {
+			yield(result.Error[[]byte](err))
```
Comment (maisem) [inline util/lineread/lineread.go:33]: I don't like this, wish you didn't have to add the type param. Oh well

File: types/result/result.go (hunk 1 above)
Comment (bradfitz) [reply types/result/result.go:15]: Seemed easier than an accessor? It's a struct value type anyway.

The v field is unexported to make people check for errors or panic. That's not needed for Err.

🤷‍♂️

Comment (maisem) [reply types/result/result.go:15]: I'd personally pick consistency here

Comment (bradfitz) [reply types/result/result.go:15]: Updated. Unexported.

File: net/netmon/interfaces_android.go [hunk 3]
```diff
@@ -54,44 +48,42 @@ func likelyHomeRouterIPAndroid() (ret netip.Addr, myIP netip.Addr, ok bool) {
...
+
 		lineNum++
 		if lineNum == 1 {
 			// Skip header line.
-			return nil
+			continue
 		}
 		if lineNum > maxProcNetRouteRead {
-			return errStopReading
+			break
```
Comment (andrew-d) [inline net/netmon/interfaces_android.go:64]: This is much nicer than `return errStopReading`

**Review by andrew-d — APPROVED** [review]
I like it!

**Review by raggi — APPROVED** [review]

**Pull request merged**